src/Compiler/Utilities/Caches.fs

```diff
@@ -0,0 +1,174 @@
+        if this.Store.Count >= options.MaximumCapacity then
+            (this.Store.Count - options.MaximumCapacity) + (options.MaximumCapacity * options.PercentageToEvict / 100)
+        else
+            0
+
+    // TODO: All of these are proofs of concept, a very naive implementation of eviction strategies, it will always walk the dictionary to find the items to evict, this is not efficient.
+    member private _.TryGetItemsToEvict () =
+        this.Store |>
+            match options.Strategy with
+            | CachingStrategy.LRU ->
+                ConcurrentDictionary.sortBy _.Value.LastAccessed
+            | CachingStrategy.LFU ->
+                ConcurrentDictionary.sortBy _.Value.AccessCount
+        |> Seq.take (this.GetEvictCount()) |> Seq.map (fun x -> x.Key)
+
+    member private _.TryEvictItems () =
+        if this.GetEvictCount() > 0 then
+            for key in this.TryGetItemsToEvict () do
+                match this.Store.TryRemove(key) with
+                | true, _ -> eviction.Trigger(key)
+                | _ -> () // TODO: We probably want to count eviction misses as well?
+
+    // TODO: Shall this be a safer task, wrapping everything in try .. with, so it's not crashing silently?
+    member private this.TryEvictTask () =
+        backgroundTask {
+            while not cts.Token.IsCancellationRequested do
+                    let evictionCount = this.GetEvictCount()
+                    if evictionCount > 0 then
+                        this.TryEvictItems ()
+                    let utilization = (this.Store.Count / options.MaximumCapacity)
+                    // So, based on utilization this will scale the delay between 0 and 1 seconds.
+                    // Worst case scenario would be when 1 second delay happens, then cache will grow rapidly, and beyond the maximum capacity.
+                    // In this case underlying dictionary will resize, AND we will have to evict items, which will likely be slow.
+                    // In this case, cache stats should be used to adjust MaximumCapacity and PercentageToEvict.
+                    let delay = 1000 - (1000 * utilization)
+                    if delay > 0 then
+                        do! Task.Delay(delay)
+
+        }
+
+    // TODO: Explore an eviction shortcut, some sort of list of keys to evict first, based on the strategy.
+    member _.TryEvict() =
+        if this.GetEvictCount() > 0 then
+            match options.EvictionMethod with
+            | EvictionMethod.Blocking -> this.TryEvictItems ()
+            | EvictionMethod.Background -> ()
+
+    member _.TryGet(key) =
+        match this.Store.TryGetValue(key) with
+        | true, value ->
+            // this is fine to be non-atomic, I guess, we are okay with race if the time is within the time of multiple concurrent calls.
+            value.LastAccessed <- DateTimeOffset.Now.Ticks
+            value.AccessCount <- Interlocked.Increment(&value.AccessCount)
```

Comment:
Why assignment? `Interlocked.Increment` already mutates?

Reply:
That's a really good question. Probably because I wasn't paying too much attention to what I was writing :(

Reply:
Fixed it